README
==============
nemo-pdf-tools
==============

A set of tools to work with PDF documents from Nemo

For more information, please visit our website:

https://github.com/atareao/nemo-pdf-tools
or
http://www.atareao.es

============
Requirements
============

Required dependencies:

python,
python-gi,
gir1.2-gtk-3.0,
gir1.2-gdkpixbuf-2.0,
gir1.2-nemo-3.0,
python-nemo,
pdf-tools

========
Download
========

Download the package:

$ git clone https://github.com/atareao/nemo-pdf-tools.git


debian/changelog
nemo-pdf-tools (0.9.0-0extras16.04.3) xenial; urgency=medium

  * splitted

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 20 Jun 2016 07:37:32 +0200

nemo-pdf-tools (0.8.0-0extras15.04.9) vivid; urgency=medium

  * Fixed minor bug

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 06 Feb 2016 11:19:21 +0100

nemo-pdf-tools (0.8.0-0extras15.04.0) vivid; urgency=medium

  * Added drag&drop feature to join pdfs
  * Added drag&drop feature to convert images to pdfs
  * Supported a lot of image formats
  * Fixed flip bug
  * Rewrited the code to separate dialogs from main code
  * Watermark with a lot of image formats

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 17 Oct 2015 21:25:23 +0200

nemo-pdf-tools (0.7.5-0extras15.04.0) vivid; urgency=medium

  * From a lot of image formats to PDF

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 17 Oct 2015 11:01:23 +0200

nemo-pdf-tools (0.7.4-0extras15.04.0) vivid; urgency=medium

  * Added new option to convert pdf to png

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 04 Oct 2015 14:01:53 +0200

nemo-pdf-tools (0.7.3-0extras15.04.0) vivid; urgency=medium

  * Added new option to create pdf file from images

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 04 Oct 2015 08:21:23 +0200

nemo-pdf-tools (0.7.2-0extras15.04.0) vivid; urgency=medium

  * Fixed quote strings
  * Added new option to reduce pdfs

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 03 Oct 2015 20:14:57 +0200

nemo-pdf-tools (0.7.1-0extras15.04.0) vivid; urgency=medium

  * Fixed extensions bug

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 27 Sep 2015 21:46:29 +0200

nemo-pdf-tools (0.7.1-0extras14.04.1) trusty; urgency=low

  * Migrated to Trusty Tahr
  * Fixed a bug with convert extensions

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 09 Mar 2014 21:23:16 +0100

nemo-pdf-tools (0.7.1-0extras13.10.3) saucy; urgency=low

  * Fixed a bug with textmark

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 30 Dec 2013 08:43:52 +0100

nemo-pdf-tools (0.7.1-0extras13.10.2) saucy; urgency=low

  * Fixed a bug with dependencies

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 29 Dec 2013 13:03:58 +0100

nemo-pdf-tools (0.7.1-0extras13.10.0) saucy; urgency=low

  * Remove PyPdf all dependencies. All manipulation is done with cairo.

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 29 Dec 2013 12:23:16 +0100

nemo-pdf-tools (0.7.0-0extras13.10.0) saucy; urgency=low

  * Added new feature to flip pdfs
  * Added a dialog to show flip and rotate operations
  * Using cairo to rotate and flip pdfs
  * Added a new feature to watermark pdfs with png images

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 28 Dec 2013 20:22:44 +0100

nemo-pdf-tools (0.6.4-0extras13.10.0) saucy; urgency=low

  * Ported to saucy

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 20 Nov 2013 12:30:23 +0100

nemo-pdf-tools (0.6.3-0extras13.04.2) raring; urgency=low

  * Ported to raring

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 02 Jun 2013 23:06:43 +0200

nemo-pdf-tools (0.0.6.2-1ubuntu1) precise; urgency=low

  * Added new paper sizes

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 14 Aug 2012 10:33:13 +0200

nemo-pdf-tools (0.0.6.1-1ubuntu1) precise; urgency=low

  * Fixed a bug with extract text feature

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 31 Mar 2012 20:44:25 +0200

nemo-pdf-tools (0.0.6.0-1ubuntu1) precise; urgency=low

  * Modified resize feature
  * Added extract text feature

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 31 Mar 2012 19:30:59 +0200

nemo-pdf-tools (0.0.5.0-1ubuntu1) precise; urgency=low

  * Added resize feature
  * Added spanish translation

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 25 Mar 2012 19:10:16 +0200

nemo-pdf-tools (0.0.4.0-1ubuntu1) precise; urgency=low

  * Added combination feature

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 25 Mar 2012 12:24:59 +0200

nemo-pdf-tools (0.0.3.1-1ubuntu1) precise; urgency=low

  * Updated spanish translation

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 17 Mar 2012 08:41:07 +0100

nemo-pdf-tools (0.0.3.0-1ubuntu1) precise; urgency=low

  * New dialog to rotate some pages
  * New dialog to extract pages from file
  * New dialog to remove pages from file
  * Now you can order files to join
  * Split one pdf file in pages

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 17 Mar 2012 08:28:30 +0100

nemo-pdf-tools (0.0.2.0-1ubuntu1) precise; urgency=low

  * Fixed a bug with spaces in filenames
  * Rotate some pages of the file
  * Extract pages of the file
  * Remove pages of the file
  * Merge several files in one

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 08 Mar 2012 06:58:49 +0100

nemo-pdf-tools (0.0.1.2-1ubuntu1) precise; urgency=low

  * Fixed a bug with special characters

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 27 Feb 2012 22:44:35 +0100

nemo-pdf-tools (0.0.1.1-1ubuntu1) precise; urgency=low

  * Fixed a bug with package

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 26 Feb 2012 20:49:00 +0100

nemo-pdf-tools (0.0.1.0-1ubuntu1) precise; urgency=low

  * First release

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 22 Jan 2012 22:00:22 +0100


src/nemo-pdf-tools.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# This file is part of nemo-pdf-tools
#
# Copyright (C) 2012-2015 Lorenzo Carbonell
# lorenzo.carbonell.cerezo@gmail.com
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import gi
try:
    gi.require_version('Nemo', '3.0')
except Exception as e:
    print(e)
    exit(1)
from gi.repository import Nemo as FileManager
from gi.repository import GObject
from gi.repository import Gtk
from gi.repository import GdkPixbuf
import os
import sys
try:
    sys.path.insert(0, '/opt/extras.ubuntu.com/pdf-tools/share/pdf-tools')
    from pdfmanager import PDFManager
    from comun import APPNAME, ICON, VERSION
    import tools
except Exception as e:
    print(e)
    exit(-1)

SEPARATOR = u'\u2015' * 10

"""
Tools to manipulate pdf
"""


class PdfToolsMenuProvider(GObject.GObject, FileManager.MenuProvider):
    """Implements the 'Replace in Filenames' extension to the File Manager
    right-click menu"""

    def __init__(self):
        """File Manager crashes if a plugin doesn't implement the __init__
        method"""
        self.pdfmanager = PDFManager()

    def doit(self, menu, option, selected):
        if option == 'rotate':
            self.pdfmanager.rotate_or_flip(selected)
        elif option == 'watermark':
            self.pdfmanager.watermark(selected)
        elif option == 'textmark':
            self.pdfmanager.textmark(selected)
        elif option == 'paginate':
            self.pdfmanager.paginate(selected)
        elif option == 'rotate pages':
            self.pdfmanager.rotate_some_pages(selected)
        elif option == 'remove pages':
            self.pdfmanager.remove_some_pages(selected)
        elif option == 'extract pages':
            self.pdfmanager.extract_some_pages(selected)
        elif option == 'join':
            self.pdfmanager.join_pdf_files(selected)
        elif option == 'split':
            self.pdfmanager.split_pdf_files(selected)
        elif option == 'combine':
            self.pdfmanager.combine_pdf_pages(selected)
        elif option == 'reduce':
            self.pdfmanager.reduce(selected)
        elif option == 'resize':
            self.pdfmanager.resize_pdf_pages(selected)
        elif option == 'convert2png':
            self.pdfmanager.convert_pdf_file_to_png(selected)
        elif option == 'convert2pdf':
            self.pdfmanager.create_pdf_from_images(selected)

    def get_file_items(self, window, sel_items):
        """Adds the 'Replace in Filenames' menu item to the File Manager
        right-click menu, connects its 'activate' signal to the 'run' method
        passing the selected Directory/File"""
        if tools.all_files_are_pdf(sel_items):
            top_menuitem = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools',
                label=_('Pdf Tools'),
                tip=_('Tools to manipulate pdf files'),
                icon='Gtk-find-and-replace')
            #
            submenu = FileManager.Menu()
            top_menuitem.set_submenu(submenu)
            sub_menus = []
            items = [
                ('01', _('Rotate and flip'), _('rotate_and_flip pdf files'),
                 'rotate'),
                ('02', _('Watermark'), _('Watermark pdffiles'),
                 'watermark'),
                ('03', _('Textmark'), _('Textmark pdf files'),
                 'textmark'),
                ('04', _('Paginate'), _('Paginate pdf files'),
                 'paginate'),
                ('05', _('Rotate pages'),
                 _('Rotate pages of the document files'),
                 'rotate pages'),
                ('06', _('Remove pages'),
                 _('Remove pages of the document files'),
                 'remove pages'),
                ('07', _('Extract pages'),
                 _('Extract pages of the document files'),
                 'extract pages'),
                ('08', _('Join pdf files'),
                 _('Join pdf files in one document'),
                 'join'),
                ('09', _('Split pdf files'),
                 _('Split a pdf in several documents'),
                 'split'),
                ('10', _('Combine pdf pages'),
                 _('Combine pdf pages in one page'),
                 'combine'),
                ('11', _('Reduce pdf size'), _('Reduce pdf size'),
                 'reduce'),
                ('12', _('Resize pdf pages'), _('Resize pdf pages'),
                 'resize'),
                ('13', _('Convert pdf to png'),
                 _('Convert pdf file to png image'),
                 'convert2png'),
            ]
            for item in items:
                sub_menuitem = FileManager.MenuItem(
                    name='PdfToolsMenuProvider::Gtk-pdf-tools-'+item[0],
                    label=item[1],
                    tip=item[2])
                sub_menuitem.connect('activate', self.doit, item[3], sel_items)
                submenu.append_item(sub_menuitem)
            #
            sub_menuitem_98 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-None',
                label=SEPARATOR)
            submenu.append_item(sub_menuitem_98)
            #
            sub_menuitem_99 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools-99',
                label=_('About'),
                tip=_('About'),
                icon='Gtk-find-and-replace')
            sub_menuitem_99.connect('activate', self.about)
            submenu.append_item(sub_menuitem_99)
            #
            return top_menuitem,
        elif tools.all_files_are_images(sel_items):
            top_menuitem = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools',
                label=_('Pdf Tools'),
                tip=_('Tools to manipulate pdf files'),
                icon='Gtk-find-and-replace')
            submenu = FileManager.Menu()
            top_menuitem.set_submenu(submenu)
            sub_menus = []
            items = [
                ('51', _('Convert to pdf'),
                 _('Convert images to pdf'),
                 'convert2pdf')
            ]
            for item in items:
                sub_menuitem = FileManager.MenuItem(
                    name='PdfToolsMenuProvider::Gtk-pdf-tools-'+item[0],
                    label=item[1],
                    tip=item[2])
                sub_menuitem.connect('activate', self.doit, item[3], sel_items)
                submenu.append_item(sub_menuitem)
            #
            sub_menuitem_98 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-None',
                label=SEPARATOR)
            submenu.append_item(sub_menuitem_98)
            #
            sub_menuitem_99 = FileManager.MenuItem(
                name='PdfToolsMenuProvider::Gtk-pdf-tools-99',
                label=_('About'),
                tip=_('About'),
                icon='Gtk-find-and-replace')
            sub_menuitem_99.connect('activate', self.about)
            submenu.append_item(sub_menuitem_99)
            #
            return top_menuitem,
        return

    def about(self, widget):
        ad = Gtk.AboutDialog()
        ad.set_name(APPNAME)
        ad.set_version(VERSION)
        ad.set_copyright('Copyrignt (c) 2011-2016\nLorenzo Carbonell')
        ad.set_comments(_('A weather indicator'))
        ad.set_license('''
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
''')
        ad.set_website('http://www.atareao.es')
        ad.set_website_label('http://www.atareao.es')
        ad.set_authors([
            'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_documenters([
            'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'])
        ad.set_logo(GdkPixbuf.Pixbuf.new_from_file(ICON))
        ad.set_icon(GdkPixbuf.Pixbuf.new_from_file(ICON))
        ad.set_program_name(APPNAME)
        ad.run()
        ad.destroy()

if __name__ == '__main__':
    '''
    cd = CreatePDFFromImagesDialog('Test')
    cd.run()
    '''
